src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs

```diff
@@ -342,6 +335,30 @@ private void UseStartup([DynamicallyAccessedMembers(StartupLinkerOptions.Accessi
                 }
             };
         });
+
+        [UnconditionalSuppressMessage("AOT", "IL3050:Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling.",
+            Justification = "There is a runtime check for ValueType startup container. It's unlikely anyone will use a ValueType here.")]
+        static void InvokeContainer(GenericWebHostBuilder genericWebHostBuilder, ConfigureContainerBuilder configureContainerBuilder)
+        {
+            var containerType = configureContainerBuilder.GetContainerType();
+
+            if (containerType.IsValueType && !RuntimeFeature.IsDynamicCodeSupported)
+            {
+                throw new InvalidOperationException("ValueType startup container isn't supported with AOT.");
```

Comment:
I don't think this is a valid approach to avoid the warning. This breaks the principle that "if you don't get any warnings, your app will work the same before and after publishing AOT".

If someone was using this functionality, and published for AOT, they wouldn't get any warnings and their app would work successfully at F5/`dotnet run` time. But when they went to publish the app, it would be broken.

Note this is different than the approach taken in https://github.com/dotnet/runtime/pull/79425. The difference is that DI is using a "feature switch" which gets enabled when `PublishAot=true` is set in the .csproj. That approach means that the app will still fail at F5/`dotnet run` time, if the app was using the AOT-incompatible functionality.

I think the existing hosting APIs should just be marked `RequiresDynamicCode` instead of suppressing these warnings.

Reply:
The same applies for the StartupLoader code/suppression.

Reply:
Startup containers are very rarely used. And I doubt anyone has created a struct startup container. I don't want to annotate the hosting APIs as RequiresDynamicCode for this one weird edge case.

Good point about `dotnet run`/`dotnet publish` difference. Should we add a flag like https://github.com/dotnet/runtime/pull/79425 does that ASP.NET Core can use to validate AOT compatibility? Or should there be one flag that is shared by libraries/applications?

Reply:
@vitek-karas @MichalStrehovsky @jkotas - thoughts on converting the DI feature switch in https://github.com/dotnet/runtime/pull/79425 to be a general purpose “Verify AOT Compatibility” feature switch that can be used by other libraries? Like for instance, here in aspnet?

Reply:
Another option is to always throw an exception for value types, AOT or not, if we think this functionality isn’t used with value types.

Reply:
Yeah, that's an option. We can publish it as a breaking change.

IMO regardless of whether we use it here or not, I think a generalized runtime verify AOT compatibility flag has some use. I think it provides a pragmatic middle ground in gray situations where something is compatible 99% of the time. Right now we have to choose between suppress and pretend everything is ok, and warn for everyone.

Reply:
> general purpose “Verify AOT Compatibility” feature switch

This sounds like https://github.com/dotnet/runtime/issues/39806

Reply:
Having something like https://github.com/dotnet/runtime/issues/39806 would be nice.

You're then left with a choice:
* Just disable the problematic case always (in our case here, always throw and potentially add class constrains in right places) - breaking change
* Rely on https://github.com/dotnet/runtime/issues/39806 to basically trigger `IsDynamicCodeSupported` based on a feature switch for you. Downside is that such solution is not granular - and so it would not be possible to disable dynamic code everywhere, except this one case.
* Introduce a new feature switch - which is what https://github.com/dotnet/runtime/pull/79425 is doing. Similar to the above, but it adds granularity. Downside is the added complexity and "yet another feature switch"
* Mark the API as AOT incompatible via the attributes - and hopefully figure out a new API which can be used instead

I think which one to do is a function of the specific scenario. If it's something super rare, I would go with either outright breaking change or generic feature switch. If it's still rare, but "real", then a dedicated feature switch might be better.